src/Generator/Passes/DelegatesPass.cs

```diff
@@ -119,61 +242,113 @@ private void AddDelegatesToDictionary(Declaration decl, Module module)
                                 }))),
                     Namespace = namespaceDelegates,
                     IsSynthetized = true,
-                    Access = decl is Method ? AccessSpecifier.Private : AccessSpecifier.Public
+                    Access = access
                 };
 
+            var typeDefType = new TypedefType
+            {
+                Declaration = @delegate
+            };
+
             var delegateString = @delegate.Visit(TypePrinter).Type;
             var existingDelegate = GetExistingDelegate(
                 module, delegateString);
 
             if (existingDelegate != null)
             {
-                Context.Delegates.Add(decl, existingDelegate);
-                return;
+                // Code to ensure that if a delegate is used for a virtual as well as something else, it finally ends up as public
+                if (existingDelegate.Access == AccessSpecifier.Private && access == AccessSpecifier.Public)
+                {
+                    existingDelegate.Access = AccessSpecifier.Public;
+                    if (libsDelegates.ContainsKey(module))
+                    {
+                        if (libsDelegates[module].ContainsKey(delegateString))
+                            libsDelegates[module].Remove(delegateString);
+                        libsDelegates[module].Add(delegateString, existingDelegate);
+                    }
+                    foreach (var dec in namespaceDelegates.Declarations.Where(m => m.Name == @delegate.Name))
+                    {
+                        namespaceDelegates.Declarations.Remove((TypedefDecl)dec);
```

Comment:
This removal looks wrong. If you find a match, you just need to make it public if not already and change the existing one, why remove anything?